### PR TITLE
Try implementing tunsocks (WIP)

### PR DESCRIPTION
**Pull request by sssemil**
Not sure what I'm doing and probably most of it is wrong. Any help would be appreciated.

https://github.com/hyperboria/bugs/issues/155

### CODE REVIEW

File: tunsocks.sh [hunk 1]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+exec /home/emil/tunsocks/tunsocks -D 8080
```
Comment (AgentME) [inline tunsocks.sh]: Maybe you should change this.

**Comment by madafoo** [comment]
Did this work? The core code of tun script implementing in openconnect is here https://github.com/chadcatlett/openconnect/blob/master/tun.c#L359. It seems your patch is very different from the original one.

File: tunsocks.sh (hunk 1 above)
Comment (sssemil) [reply tunsocks.sh]: Ah, right. Thanks, I'll remove it.

**Comment by sssemil** [comment]
Nope. I was trying to use libuv here, probably that's why it's so different.

**Comment by sssemil** [comment]
To make it work I need to start tunsocks from cjdroute and stream stdio from cjdroute's interface thing to tunsocks stdio (as far as I know if it's even right).

**Comment by kpcyrd** [comment]
@sssemil that's how I understood it as well from the [readme](https://github.com/russdill/tunsocks).

Instead of opening /dev/net/tun for read/write, cjdroute would have to execute tunsocks and then read/write from stdin/stdout. You might run into issues with cjdns' sandbox if you try to exec a binary. If that's the case, you could try writing a tun adapter that expects a pseudo tun device on specific fds that are passed to cjdroute by a wrapper.

**Comment by sssemil** [comment]
@kpcyrd like this? :
```
#include "library.h"

#include <stdio.h>

#define __USE_GNU

#include <dlfcn.h>
#include <string.h>

#if defined(android)
#define DEVICE_PATH "/dev/tun"
#else
#define DEVICE_PATH "/dev/net/tun"
#endif

typedef int (*orig_ioctl_f_type)(int d, unsigned long request, char *argp);

int ioctl(int d, unsigned long request, char *argp, ...) {
    orig_ioctl_f_type orig_ioctl;
    orig_ioctl = (orig_ioctl_f_type) dlsym(RTLD_NEXT, "ioctl");
    int tmp = orig_ioctl(d, request, argp);

    if (d == 1001) {
        printf("Hello from ioctl: %d, %d\n", d, tmp);
    }

    return tmp;
}

typedef int (*orig_open_f_type)(const char *pathname, int flags);

int open(const char *pathname, int flags, ...) {
    if (strcmp(pathname, DEVICE_PATH) == 0){
        printf("Hello from open: \"%s\"\n", pathname);
        return 1001;
    }

    orig_open_f_type orig_open;
    orig_open = (orig_open_f_type) dlsym(RTLD_NEXT, "open");
    return orig_open(pathname, flags);
}
```

**Comment by kylerchin** [comment]
Hi @sssemil @kpcyrd @cjdelisle! Are ya'll still working on this or do you want to merge?

**Comment by sssemil** [comment]
Hi @kylerchin, it's not ready yet. I'll get back to it March. I'll close it till then.

**Pull request closed without merging**